hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/SparkMetadataTableRecordIndex.java

```diff
@@ -0,0 +1,227 @@
+    // Only those deltacommits which have a valid completed commit on the dataset are read. Since, the instantTime
+    // is being rolled back on the dataset, we will not load the records from the deltacommit and it is virtually
+    // rolled back. In other words, there is no need to rollback any deltacommit here except if the deltacommit
+    // was compacted and a new basefile has been created.
+    try {
+      HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder()
+          .setBasePath(HoodieTableMetadata.getMetadataTableBasePath(config.getBasePath()))
+          .setConf(new Configuration()).build();
+      HoodieTimeline commitTimeline = metaClient.getCommitTimeline().filterCompletedInstants();
+      if (commitTimeline.empty()) {
+        // No compaction yet so no need to check for deltacommits due to the logic above
+        return true;
+      }
+
+      if (HoodieTimeline.compareTimestamps(instantTime, GREATER_THAN, commitTimeline.lastInstant().get().getTimestamp())) {
+        // After the last compaction so no rollback required as per logic above
+        return true;
+      }
+      LOG.warn("Cannot rollback instant " + instantTime + " because the corresponding deltacommit has been compacted "
+          + " in " + commitTimeline.lastInstant().get().getTimestamp());
+      return false;
+    } catch (TableNotFoundException e) {
+      // Metadata table is not setup.  Nothing to rollback.  Exit gracefully.
+      LOG.warn("Cannot rollback instant " + instantTime + " as metadata table is not found");
+      return true;
+    }
+  }
+
+  @Override
+  public boolean isGlobal() {
+    return true;
+  }
+
+  @Override
+  public boolean canIndexLogFiles() {
+    return true;
+  }
+
+  @Override
+  public boolean isImplicitWithStorage() {
+    return false;
+  }
+
+  private <R> HoodieData<HoodieRecord<R>> tagLocationBackToRecords(
+      HoodiePairData<String, HoodieRecordGlobalLocation> keyFilenamePair,
+      HoodieData<HoodieRecord<R>> records) {
+    HoodiePairData<String, HoodieRecord<R>> keyRecordPairs =
+        records.mapToPair(record -> new ImmutablePair<>(record.getRecordKey(), record));
+    // Here as the records might have more data than keyFilenamePairs (some row keys' not found in record index),
+    // we will do left outer join.
+    return keyRecordPairs.leftOuterJoin(keyFilenamePair).values()
+        .map(v -> {
+          HoodieRecord<R> record = v.getLeft();
+          Option<HoodieRecordGlobalLocation> location = Option.ofNullable(v.getRight().orElse(null));
+          if (!location.isPresent()) {
+            // No location found.
+            return record;
+          }
+          // Ensure the partitionPath is also set correctly in the key
+          if (!record.getPartitionPath().equals(location.get().getPartitionPath())) {
+            record = new HoodieAvroRecord(new HoodieKey(record.getRecordKey(), location.get().getPartitionPath()), (HoodieRecordPayload) record.getData());
```

Comment:
Does this work of the record type is Spark?  
may be we can invoke this method in HoodieIndexUtils
```
  public static <R> HoodieRecord<R> getTaggedRecord(HoodieRecord<R> inputRecord, Option<HoodieRecordLocation> location) {
```

Reply:
it should check `config.getRecordMerger().getRecordType()` and create record accordingly

Reply:
Corrected the creation of new record.

HoodieIndexUtils. getTaggedRecord cannot be used as it expects HoodieRecordLocation as second argument and in the function we have HoodieRecordGlobalLocation.